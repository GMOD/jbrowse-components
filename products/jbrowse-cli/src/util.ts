import fs from 'fs'
import path from 'path'
import { isURL, createRemoteStream } from './types/common'

export async function getLocalOrRemoteStream(uri: string, out: string) {
  if (isURL(uri)) {
    const result = await createRemoteStream(uri)
    return {
      totalBytes: +(result.headers.get('Content-Length') || 0),
      stream: result.body,
    }
  } else {
    const filename = path.isAbsolute(uri) ? uri : path.join(out, uri)
    return {
      totalBytes: fs.statSync(filename).size,
      stream: fs.createReadStream(filename),
    }
  }
}

export function decodeURIComponentNoThrow(uri: string) {
  try {
    return decodeURIComponent(uri)
  } catch (e) {
    // avoid throwing exception on a failure to decode URI component
    return uri
  }
}
