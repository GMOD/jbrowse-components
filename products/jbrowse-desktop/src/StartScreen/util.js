import PluginManager from '@jbrowse/core/PluginManager'
import PluginLoader from '@jbrowse/core/PluginLoader'
import { readConfObject } from '@jbrowse/core/configuration'

import JBrowseRootModelFactory from '../rootModel'
import corePlugins from '../corePlugins'
import { version } from '../../package.json'

import {
  writeAWSAnalytics,
  writeGAAnalytics,
} from '@jbrowse/core/util/analytics'

export async function createPluginManager(
  configSnapshot: any,
  initialTimestamp = +Date.now(),
) {
  const pluginLoader = new PluginLoader(configSnapshot.plugins)
  pluginLoader.installGlobalReExports(window)
  const runtimePlugins = await pluginLoader.load()
  const plugins = [
    ...corePlugins.map(P => ({
      plugin: new P(),
      metadata: { isCore: true },
    })),
    ...runtimePlugins.map(({ plugin: P, definition }) => ({
      plugin: new P(),
      definition,
      metadata: { url: definition.url },
    })),
  ]
  const pm = new PluginManager(plugins)
  pm.createPluggableElements()

  const JBrowseRootModel = JBrowseRootModelFactory(pm)
  const rootModel = JBrowseRootModel.create(
    {
      jbrowse: configSnapshot,
      assemblyManager: {},
      version,
    },
    { pluginManager: pm },
  )

  const config = rootModel.jbrowse.configuration
  const { rpc } = config

  rpc.addDriverConfig('WebWorkerRpcDriver', {
    type: 'WebWorkerRpcDriver',
  })
  rpc.defaultDriver.set('WebWorkerRpcDriver')

  pm.setRootModel(rootModel)
  pm.configure()

  if (rootModel && !readConfObject(config, 'disableAnalytics')) {
    writeAWSAnalytics(rootModel, initialTimestamp)
    writeGAAnalytics(rootModel, initialTimestamp)
  }

  rootModel.setDefaultSession()

  return pm
}
