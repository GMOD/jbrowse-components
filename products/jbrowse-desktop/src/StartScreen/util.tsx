import { useState, useEffect } from 'react'
import PluginManager from '@jbrowse/core/PluginManager'
import PluginLoader from '@jbrowse/core/PluginLoader'
import { readConfObject } from '@jbrowse/core/configuration'
import deepmerge from 'deepmerge'
import { ipcRenderer } from 'electron'

import JBrowseRootModelFactory from '../rootModel'
import corePlugins from '../corePlugins'
import { version } from '../../package.json'

import {
  writeAWSAnalytics,
  writeGAAnalytics,
} from '@jbrowse/core/util/analytics'

const defaultInternetAccounts = [
  {
    type: 'DropboxOAuthInternetAccount',
    internetAccountId: 'dropboxOAuth',
    name: 'Dropbox',
    description: 'OAuth Info for Dropbox',
    authEndpoint: 'https://www.dropbox.com/oauth2/authorize',
    tokenEndpoint: 'https://api.dropbox.com/oauth2/token',
    needsAuthorization: true,
    needsPKCE: true,
    hasRefreshToken: true,
    clientId: 'ykjqg1kr23pl1i7',
    domains: [
      'addtodropbox.com',
      'db.tt',
      'dropbox.com',
      'dropboxapi.com',
      'dropboxbusiness.com',
      'dropbox.tech',
      'getdropbox.com',
    ],
  },
  {
    type: 'GoogleDriveOAuthInternetAccount',
    internetAccountId: 'googleOAuth',
    name: 'Google',
    description: 'OAuth Info for Google Drive',
    authEndpoint: 'https://accounts.google.com/o/oauth2/v2/auth',
    needsAuthorization: true,
    clientId:
      '109518325434-m86s8a5og8ijc5m6n7n8dk7e9586bg9i.apps.googleusercontent.com',
    scopes: 'https://www.googleapis.com/auth/drive.readonly',
    responseType: 'token',
    domains: ['drive.google.com'],
  },
]

export async function loadPluginManager(filePath: string) {
  const snap = await ipcRenderer.invoke('loadSession', filePath)
  const pm = await createPluginManager(snap)
  // @ts-ignore
  pm.rootModel?.setSessionPath(filePath)
  return pm
}

export async function createPluginManager(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  configSnapshot: any,
  initialTimestamp = +Date.now(),
) {
  const pluginLoader = new PluginLoader(configSnapshot.plugins)
  pluginLoader.installGlobalReExports(window)
  const runtimePlugins = await pluginLoader.load()
  const pm = new PluginManager([
    ...corePlugins.map(P => ({
      plugin: new P(),
      metadata: { isCore: true },
    })),
    ...runtimePlugins.map(({ plugin: P, definition }) => ({
      plugin: new P(),
      definition,
      metadata: { url: definition.url },
    })),
  ])
  pm.createPluggableElements()

  const JBrowseRootModel = JBrowseRootModelFactory(pm)

  const jbrowse = deepmerge(configSnapshot, {
    internetAccounts: defaultInternetAccounts,
    assemblies: [],
    tracks: [],
  }) as {
    internetAccounts: { internetAccountId: string }[]
    assemblies: { name: string }[]
    tracks: { trackId: string }[]
  }

  const internetAccountIds = jbrowse.internetAccounts.map(
    o => o.internetAccountId,
  )

  // remove duplicates while mixing in default internet accounts
  jbrowse.internetAccounts = jbrowse.internetAccounts.filter(
    ({ internetAccountId }, index) =>
      !internetAccountIds.includes(internetAccountId, index + 1),
  )

  const assemblyNames = jbrowse.assemblies.map(o => o.name)

  jbrowse.assemblies = jbrowse.assemblies.filter(
    ({ name }, index) => !assemblyNames.includes(name, index + 1),
  )

  const trackIds = jbrowse.tracks.map(o => o.trackId)

  jbrowse.tracks = jbrowse.tracks.filter(
    ({ trackId }, index) => !trackIds.includes(trackId, index + 1),
  )

  const rootModel = JBrowseRootModel.create(
    {
      jbrowse,
      assemblyManager: {},
      version,
    },
    { pluginManager: pm },
  )

  const config = rootModel.jbrowse.configuration
  const { rpc } = config

  rpc.addDriverConfig('WebWorkerRpcDriver', {
    type: 'WebWorkerRpcDriver',
  })
  rpc.defaultDriver.set('WebWorkerRpcDriver')

  pm.setRootModel(rootModel)
  pm.configure()

  if (rootModel && !readConfObject(config, 'disableAnalytics')) {
    writeAWSAnalytics(rootModel, initialTimestamp)
    writeGAAnalytics(rootModel, initialTimestamp)
  }

  rootModel.setDefaultSession()

  return pm
}

// similar to https://blog.logrocket.com/using-localstorage-react-hooks/
export const useLocalStorage = (key: string, defaultValue: string) => {
  const [value, setValue] = useState(
    () => localStorage.getItem(key) || defaultValue,
  )

  useEffect(() => {
    localStorage.setItem(key, value)
  }, [key, value])

  // without this cast, tsc complained that the type of setValue could be a
  // string or a callback
  return [value, setValue] as [string, (arg: string) => void]
}
