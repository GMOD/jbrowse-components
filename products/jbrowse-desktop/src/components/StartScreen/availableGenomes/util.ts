import { useEffect, useState } from 'react'

export function useInnerDims() {
  const [height, setHeight] = useState(window.innerHeight)
  const [width, setWidth] = useState(window.innerWidth)

  useEffect(() => {
    const updateWindowDimensions = () => {
      setWidth(window.innerWidth)
      setHeight(window.innerHeight)
    }

    window.addEventListener('resize', updateWindowDimensions)

    return () => {
      window.removeEventListener('resize', updateWindowDimensions)
    }
  }, [])
  return {
    height,
    width,
  }
}
