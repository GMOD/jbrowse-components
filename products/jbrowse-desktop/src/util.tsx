import fsPromises from 'fs/promises'
import os from 'os'
import path from 'path'

import sanitize from 'sanitize-filename'
import type { LoadedPlugin } from '@jbrowse/core/PluginLoader'

export async function fetchCJS(url: string): Promise<LoadedPlugin> {
  // On macOS `os.tmpdir()` returns the path to a symlink, see:
  // https://github.com/nodejs/node/issues/11422
  const tmpDir = await fsPromises.mkdtemp(
    path.join(await fsPromises.realpath(os.tmpdir()), 'jbrowse-plugin-'),
  )
  try {
    const pluginLocation = path.join(tmpDir, sanitize(url))
    const pluginLocationRelative = path.relative('.', pluginLocation)
    const response = await fetch(url)
    if (!response.ok) {
      throw new Error(
        `HTTP ${response.status} ${response.statusText} when fetching plugin: ${url})`,
      )
    }
    const pluginText = await response.text()
    await fsPromises.writeFile(pluginLocation, pluginText)
    return globalThis.require(pluginLocationRelative) as LoadedPlugin
  } finally {
    await fsPromises.rmdir(tmpDir, { recursive: true })
  }
}
