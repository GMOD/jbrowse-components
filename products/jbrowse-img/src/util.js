// nice helper function from https://stackoverflow.com/questions/263965/
export const booleanize = (string) => (string === "false" ? false : !!string);
