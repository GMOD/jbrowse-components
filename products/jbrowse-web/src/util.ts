import {
  PluginDefinition,
  isCJSPluginDefinition,
  isESMPluginDefinition,
  isUMDPluginDefinition,
} from '@jbrowse/core/PluginLoader'
import { AnyConfigurationModel } from '@jbrowse/core/configuration'
import {
  getPropertyMembers,
  getChildType,
  isArrayType,
  isModelType,
  isReferenceType,
  isValidReference,
  isMapType,
  types,
  IAnyType,
  IAnyStateTreeNode,
  Instance,
  SnapshotOut,
} from 'mobx-state-tree'

/**
 * Pad the end of a base64 string with "=" to make it valid
 * @param b64 - unpadded b64 string
 */
export function b64PadSuffix(b64: string): string {
  let num = 0
  const mo = b64.length % 4
  switch (mo) {
    case 3:
      num = 1
      break
    case 2:
      num = 2
      break
    case 0:
      num = 0
      break
    default:
      throw new Error('base64 not a valid length')
  }
  return b64 + '='.repeat(num)
}

/**
 * Decode and inflate a url-safe base64 to a string
 * See {@link https://en.wikipedia.org/wiki/Base64#URL_applications}
 * @param b64 - a base64 string to decode and inflate
 */
export async function fromUrlSafeB64(b64: string) {
  const originalB64 = b64PadSuffix(
    b64.replaceAll('-', '+').replaceAll('_', '/'),
  )
  const { toByteArray } = await import('base64-js')
  const { inflate } = await import('pako')
  const bytes = toByteArray(originalB64)
  const inflated = inflate(bytes)
  return new TextDecoder().decode(inflated)
}

/**
 * Compress and encode a string as url-safe base64
 * See {@link https://en.wikipedia.org/wiki/Base64#URL_applications}
 * @param str-  a string to compress and encode
 */
export async function toUrlSafeB64(str: string) {
  const bytes = new TextEncoder().encode(str)
  const { deflate } = await import('pako')
  const { fromByteArray } = await import('base64-js')
  const deflated = deflate(bytes)
  const encoded = fromByteArray(deflated)
  const pos = encoded.indexOf('=')
  return pos > 0
    ? encoded.slice(0, pos).replaceAll('+', '-').replaceAll('/', '_')
    : encoded.replaceAll('+', '-').replaceAll('/', '_')
}

type MSTArray = Instance<ReturnType<typeof types.array>>
type MSTMap = Instance<ReturnType<typeof types.map>>

// attempts to remove undefined references from the given MST model. can only actually
// remove them from arrays and maps. throws MST undefined ref error if it encounters
// undefined refs in model properties
export function filterSessionInPlace(node: IAnyStateTreeNode, type: IAnyType) {
  // makes it work with session sharing
  if (node === undefined) {
    return
  }
  if (isArrayType(type)) {
    const array = node as MSTArray
    const childType = getChildType(node)
    if (isReferenceType(childType)) {
      // filter array elements
      for (let i = 0; i < array.length; ) {
        if (!isValidReference(() => array[i])) {
          array.splice(i, 1)
        } else {
          i += 1
        }
      }
    }
    array.forEach(el => {
      filterSessionInPlace(el, childType)
    })
  } else if (isMapType(type)) {
    const map = node as MSTMap
    const childType = getChildType(map)
    if (isReferenceType(childType)) {
      // filter the map members
      for (const key in map.keys()) {
        if (!isValidReference(() => map.get(key))) {
          map.delete(key)
        }
      }
    }
    map.forEach(child => {
      filterSessionInPlace(child, childType)
    })
  } else if (isModelType(type)) {
    // iterate over children
    const { properties } = getPropertyMembers(node)

    Object.entries(properties).forEach(([pname, ptype]) => {
      // @ts-ignore
      filterSessionInPlace(node[pname], ptype)
    })
  }
}

type Config = SnapshotOut<AnyConfigurationModel>

export function addRelativeUris(config: Config, base: URL) {
  if (typeof config === 'object') {
    for (const key of Object.keys(config)) {
      if (typeof config[key] === 'object') {
        addRelativeUris(config[key], base)
      } else if (key === 'uri') {
        config.baseUri = base.href
      }
    }
  }
}

type Root = { configuration?: Config }

// raw readConf alternative for before conf is initialized
export function readConf({ configuration }: Root, attr: string, def: string) {
  return configuration?.[attr] || def
}

export async function fetchPlugins() {
  const response = await fetch('https://jbrowse.org/plugin-store/plugins.json')
  if (!response.ok) {
    throw new Error(
      `HTTP ${response.status} ${response.statusText} fetching plugins`,
    )
  }
  return response.json() as Promise<{ plugins: PluginDefinition[] }>
}

export async function checkPlugins(pluginsToCheck: PluginDefinition[]) {
  if (pluginsToCheck.length === 0) {
    return true
  }
  const storePlugins = await fetchPlugins()
  return pluginsToCheck.every(p => {
    if (isUMDPluginDefinition(p)) {
      return storePlugins.plugins.some(
        pp =>
          isUMDPluginDefinition(p) &&
          (('url' in pp && 'url' in p && p.url === pp.url) ||
            ('umdUrl' in pp && 'umdUrl' in p && p.umdUrl === pp.umdUrl)),
      )
    }
    if (isESMPluginDefinition(p)) {
      return storePlugins.plugins.some(
        pp =>
          isESMPluginDefinition(p) && 'esmUrl' in p && p.esmUrl === pp.esmUrl,
      )
    }
    if (isCJSPluginDefinition(p)) {
      return storePlugins.plugins.some(
        pp => isCJSPluginDefinition(p) && p.cjsUrl === pp.cjsUrl,
      )
    }
    return false
  })
}

export function removeAttr(obj: Record<string, unknown>, attr: string) {
  for (const prop in obj) {
    if (prop === attr) {
      delete obj[prop]
    } else if (typeof obj[prop] === 'object') {
      removeAttr(obj[prop] as Record<string, unknown>, attr)
    }
  }
  return obj
}
