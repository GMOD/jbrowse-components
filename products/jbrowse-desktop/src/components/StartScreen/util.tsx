import PluginLoader from '@jbrowse/core/PluginLoader'
import PluginManager from '@jbrowse/core/PluginManager'
import { readConfObject } from '@jbrowse/core/configuration'
import {
  writeAWSAnalytics,
  writeGAAnalytics,
} from '@jbrowse/core/util/analytics'
import deepmerge from 'deepmerge'

// locals
import corePlugins from '../../corePlugins'
import JBrowseRootModelFactory from '../../rootModel'
import sessionModelFactory from '../../sessionModel'
import { fetchCJS } from '../../util'
import type { PluginDefinition } from '@jbrowse/core/PluginLoader'

const { ipcRenderer } = window.require('electron')

function uniqBy<T>(a: T[], key: (arg: T) => string) {
  const seen = new Set()
  return a.filter(item => {
    const k = key(item)
    return seen.has(k) ? false : seen.add(k)
  })
}

const defaultInternetAccounts = [
  {
    type: 'DropboxOAuthInternetAccount',
    internetAccountId: 'dropboxOAuth',
    name: 'Dropbox',
    description: 'Account to access Dropbox files',
    clientId: 'ykjqg1kr23pl1i7',
  },
  {
    type: 'GoogleDriveOAuthInternetAccount',
    internetAccountId: 'googleOAuth',
    name: 'Google Drive',
    description: 'Account to access Google Drive files',
    clientId:
      '109518325434-m86s8a5og8ijc5m6n7n8dk7e9586bg9i.apps.googleusercontent.com',
  },
]

export async function loadPluginManager(configPath: string) {
  const snap = await ipcRenderer.invoke('loadSession', configPath)
  const pm = await createPluginManager(snap)
  // @ts-expect-error
  pm.rootModel?.setSessionPath(configPath)
  return pm
}

export async function createPluginManager(
  configSnapshot: {
    plugins?: PluginDefinition[]
  },
  initialTimestamp = +Date.now(),
) {
  const pluginLoader = new PluginLoader(configSnapshot.plugins, {
    fetchESM: url => import(/* webpackIgnore:true */ url),
    fetchCJS,
  })
  pluginLoader.installGlobalReExports(window)
  const runtimePlugins = await pluginLoader.load(window.location.href)
  const pluginManager = new PluginManager([
    ...corePlugins.map(P => ({
      plugin: new P(),
      metadata: {
        isCore: true,
      },
    })),
    ...runtimePlugins.map(({ plugin: P, definition }) => ({
      plugin: new P(),
      definition,
      metadata: {
        // @ts-expect-error
        url: definition.url,
        // @ts-expect-error
        esmUrl: definition.esmUrl,
        // @ts-expect-error
        umdUrl: definition.umdUrl,
        // @ts-expect-error
        cjsUrl: definition.cjsUrl,
      },
    })),
  ])
  pluginManager.createPluggableElements()

  const JBrowseRootModel = JBrowseRootModelFactory({
    pluginManager,
    sessionModelFactory,
  })

  const jbrowse = deepmerge(configSnapshot, {
    internetAccounts: defaultInternetAccounts,
    assemblies: [],
    tracks: [],
  }) as {
    internetAccounts: { internetAccountId: string }[]
    assemblies: { name: string }[]
    tracks: { trackId: string }[]
  }

  jbrowse.assemblies = uniqBy(jbrowse.assemblies, asm => asm.name)
  jbrowse.tracks = uniqBy(jbrowse.tracks, acct => acct.trackId)
  jbrowse.internetAccounts = uniqBy(
    jbrowse.internetAccounts,
    acct => acct.internetAccountId,
  )

  const rootModel = JBrowseRootModel.create({ jbrowse }, { pluginManager })
  const config = rootModel.jbrowse.configuration
  const { rpc } = config
  rpc.defaultDriver.set('WebWorkerRpcDriver')

  pluginManager.setRootModel(rootModel)
  pluginManager.configure()

  if (!readConfObject(config, 'disableAnalytics')) {
    // these are ok if they are uncaught promises
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    writeAWSAnalytics(rootModel, initialTimestamp)
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    writeGAAnalytics(rootModel, initialTimestamp)
  }

  rootModel.setDefaultSession()

  return pluginManager
}

export interface RecentSessionData {
  path: string
  name: string
  screenshot?: string
  updated?: number
}
