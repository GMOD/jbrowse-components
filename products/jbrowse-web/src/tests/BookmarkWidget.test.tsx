import { waitFor, fireEvent, within } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { saveAs } from 'file-saver'
import { createView, setup, doBeforeEach } from './util'

jest.mock('file-saver', () => {
  return {
    ...jest.requireActual('file-saver'),
    saveAs: jest.fn(),
  }
})
setup()

beforeEach(() => {
  doBeforeEach()
  localStorage.clear()
})

const delay = { timeout: 30000 }
const opts = [{}, delay]

test('Open the bookmarks widget from the top level menu', async () => {
  const { findByText } = await createView()

  const user = userEvent.setup()
  await user.click(await findByText('Tools'))
  await user.click(await findByText('Bookmarks'))

  expect(await findByText('Bookmarked regions')).toBeTruthy()
})

test('Open the bookmarks widget from the view menu', async () => {
  const { findByTestId, findByText } = await createView()

  const user = userEvent.setup()
  await user.click(await findByTestId('view_menu_icon'))
  await user.click(await findByText('Open bookmark widget'))

  expect(await findByText('Bookmarked regions')).toBeTruthy()
})

test('Create a bookmark using the click and drag rubberband', async () => {
  const { session, findByTestId, findByText } = await createView()
  const rubberband = await findByTestId('rubberband_controls', {}, delay)

  fireEvent.mouseDown(rubberband, { clientX: 100, clientY: 0 })
  fireEvent.mouseMove(rubberband, { clientX: 250, clientY: 0 })
  fireEvent.mouseUp(rubberband, { clientX: 250, clientY: 0 })
  fireEvent.click(await findByText('Bookmark region'))

  // @ts-expect-error
  const { bookmarks } = session.widgets.get('GridBookmark')
  expect(bookmarks).toMatchSnapshot()
}, 40000)

test('Create a bookmark using the hotkey to bookmark the current region', async () => {
  const { session, findByTestId } = await createView()

  const user = userEvent.setup()
  await user.click(await findByTestId('trackContainer', ...opts))

  document.dispatchEvent(
    new KeyboardEvent('keydown', {
      code: 'KeyD',
      shiftKey: true,
      ctrlKey: true,
    }),
  )

  // @ts-expect-error
  const { bookmarks } = session.widgets.get('GridBookmark')
  expect(bookmarks).toMatchSnapshot()
})

test('Create a bookmark using the menu button to bookmark the current region', async () => {
  const { session, findByTestId, findByText } = await createView()

  const user = userEvent.setup()
  await user.click(await findByTestId('trackContainer', ...opts))
  await user.click(await findByTestId('view_menu_icon'))
  await user.click(await findByText('Bookmarks'))
  await user.click(await findByText('Bookmark current region'))

  // @ts-expect-error
  const { bookmarks } = session.widgets.get('GridBookmark')
  expect(bookmarks).toMatchSnapshot()
}, 40000)

test('Navigate to a bookmark using the embedded link in the widget data grid', async () => {
  const { view, session, findByTestId, findByText } = await createView()

  const user = userEvent.setup()
  await user.click(await findByTestId('view_menu_icon'))
  await user.click(await findByText('Open bookmark widget'))

  // @ts-expect-error
  const bookmarkWidget = session.widgets.get('GridBookmark')
  bookmarkWidget.addBookmark({
    start: 200,
    end: 240,
    refName: 'ctgA',
    assemblyName: 'volvox',
  })

  fireEvent.click(await findByText('ctgA:201..240', {}, delay))
  await waitFor(() => {
    expect(view.visibleLocStrings).toBe('ctgA:201..240')
  })
}, 40000)

test('Navigate to a bookmark using the hotkey to navigate to the most recently created bookmark', async () => {
  const { view, session, findByTestId, findByText } = await createView()

  const user = userEvent.setup()
  await user.click(await findByTestId('view_menu_icon'))
  await user.click(await findByText('Bookmarks'))
  await user.click(await findByText('Open bookmark widget'))

  // @ts-expect-error
  const bookmarkWidget = session.widgets.get('GridBookmark')
  bookmarkWidget.addBookmark({
    start: 200,
    end: 240,
    refName: 'ctgA',
    assemblyName: 'volvox',
  })

  document.dispatchEvent(
    new KeyboardEvent('keydown', {
      code: 'KeyM',
      shiftKey: true,
      ctrlKey: true,
    }),
  )

  await waitFor(() => {
    expect(view.visibleLocStrings).toBe('ctgA:201..240')
  })
}, 40000)

test('Edit a bookmark label with a single click on the data grid', async () => {
  const { session, findByText, findAllByRole } = await createView()

  const user = userEvent.setup()
  await user.click(await findByText('Tools'))
  await user.click(await findByText('Bookmarks'))

  // @ts-expect-error
  const bookmarkWidget = session.widgets.get('GridBookmark')
  bookmarkWidget.addBookmark({
    start: 200,
    end: 240,
    refName: 'ctgA',
    assemblyName: 'volvox',
  })

  const field = (await findAllByRole('gridcell'))[2]!
  await user.type(field, 'new label')
  // get the focus away from the field
  fireEvent.click(document)

  expect(field.innerHTML).toContain('new label')
})
test('Edit a bookmark label with a double click via the dialog', async () => {
  const { session, findByText, findAllByRole, findByTestId } =
    await createView()

  const user = userEvent.setup()
  await user.click(await findByText('Tools'))
  await user.click(await findByText('Bookmarks'))

  // @ts-expect-error
  const bookmarkWidget = session.widgets.get('GridBookmark')
  bookmarkWidget.addBookmark({
    start: 200,
    end: 240,
    refName: 'ctgA',
    assemblyName: 'volvox',
  })

  const field = (await findAllByRole('gridcell'))[2]!

  await user.dblClick(field)
  await user.type(await findByTestId('edit-bookmark-label-field'), 'new label')
  await user.click(await findByText('Confirm'))
  expect(field.innerHTML).toContain('new label')
})

test('Toggle highlight visibility across all views', async () => {
  const { session, findByText, findByTestId, findAllByTestId } =
    await createView()

  session.addView('LinearGenomeView', {
    displayedRegions: [
      {
        assemblyName: 'volvox',
        refName: 'ctgA',
        start: 0,
        end: 1000,
      },
    ],
  })

  const user = userEvent.setup()

  await user.click(await findByText('Tools'))
  await user.click(await findByText('Bookmarks'))

  // @ts-expect-error
  const bookmarkWidget = session.widgets.get('GridBookmark')
  bookmarkWidget.addBookmark({
    start: 200,
    end: 240,
    refName: 'ctgA',
    assemblyName: 'volvox',
  })

  const highlight = (await findAllByTestId('BookmarkIcon'))[0]
  const highlight2 = (await findAllByTestId('BookmarkIcon'))[1]

  expect(highlight).toBeDefined()
  expect(highlight2).toBeDefined()

  fireEvent.click(await findByTestId('grid_bookmark_menu', ...opts))
  await user.click(await findByText('Settings'))
  await user.click(await findByTestId('toggle_highlight_all_switch'))
  await user.click(await findByText('Close'))

  // expect(highlight3).toBeUndefined()
  // expect(highlight4).toBeUndefined()
})

test('Toggle highlight label visibility across all views', async () => {
  const { session, findByText, findByTestId } = await createView()

  const user = userEvent.setup()

  await user.click(await findByText('Tools'))
  await user.click(await findByText('Bookmarks'))

  // @ts-expect-error
  const bookmarkWidget = session.widgets.get('GridBookmark')
  bookmarkWidget.addBookmark({
    start: 200,
    end: 240,
    refName: 'ctgA',
    assemblyName: 'volvox',
  })

  // const highlight = (await findAllByTestId('BookmarkIcon'))[0]
  // const highlight2 = (await findAllByTestId('BookmarkIcon'))[1]

  // expect(highlight).toBeDefined()
  // expect(highlight2).toBeDefined()

  fireEvent.click(await findByTestId('grid_bookmark_menu', ...opts))
  await user.click(await findByText('Settings'))
  await user.click(await findByTestId('toggle_highlight_label_all_switch'))
  await user.click(await findByText('Close'))

  // expect(highlight).toBeUndefined()
  // expect(highlight2).toBeUndefined()
})

test('Downloads a BED file correctly', async () => {
  const { session, findByText, findByTestId } = await createView()

  // @ts-expect-error
  const bookmarkWidget = session.addWidget(
    'GridBookmarkWidget',
    'gridBookmarkWidget',
  )
  // @ts-expect-error
  session.showWidget('gridBookmarkWidget')

  bookmarkWidget.addBookmark({
    refName: 'ctgA',
    start: 0,
    end: 8,
    assemblyName: 'volvox',
  })

  fireEvent.click(await findByTestId('grid_bookmark_menu', ...opts))
  fireEvent.click(await findByText('Export', ...opts))
  fireEvent.click(await findByText(/Download/, ...opts))

  const blob = new Blob([''], {
    type: 'text/x-bed;charset=utf-8',
  })

  expect(saveAs).toHaveBeenCalledWith(blob, 'jbrowse_bookmarks_volvox.bed')
}, 20000)

test('Downloads a TSV file correctly', async () => {
  const { session, findByText, findByTestId, getByRole } = await createView()

  // @ts-expect-error
  const bookmarkWidget = session.addWidget(
    'GridBookmarkWidget',
    'gridBookmarkWidget',
  )
  // @ts-expect-error
  session.showWidget('gridBookmarkWidget')

  bookmarkWidget.addBookmark({
    refName: 'ctgA',
    start: 0,
    end: 8,
    assemblyName: 'volvox',
  })
  fireEvent.click(await findByTestId('grid_bookmark_menu'))
  fireEvent.click(await findByText('Export'))
  fireEvent.mouseDown(await findByText('BED'))
  const listbox = within(getByRole('listbox'))
  fireEvent.click(listbox.getByText('TSV'))
  fireEvent.click(await findByText(/Download/))

  const blob = new Blob([''], {
    type: 'text/tab-separated-values;charset=utf-8',
  })

  expect(saveAs).toHaveBeenCalledWith(blob, 'jbrowse_bookmarks.tsv')
})
