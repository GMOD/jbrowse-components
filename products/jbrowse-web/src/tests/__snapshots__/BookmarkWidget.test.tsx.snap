// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create a bookmark using the click and drag rubberband 1`] = `
[
  {
    "assemblyName": "volvox",
    "end": 113,
    "highlight": "rgba(247, 129, 192, 0.35)",
    "label": "",
    "refName": "ctgA",
    "reversed": false,
    "start": 105,
  },
]
`;

exports[`Create a bookmark using the hotkey to bookmark the current region 1`] = `
[
  {
    "assemblyName": "volvox",
    "end": 140,
    "highlight": "rgba(247, 129, 192, 0.35)",
    "label": "",
    "refName": "ctgA",
    "reversed": false,
    "start": 100,
  },
]
`;

exports[`Create a bookmark using the menu button to bookmark the current region 1`] = `
[
  {
    "assemblyName": "volvox",
    "end": 140,
    "highlight": "rgba(247, 129, 192, 0.35)",
    "label": "",
    "refName": "ctgA",
    "reversed": false,
    "start": 100,
  },
]
`;
